### PR TITLE
feat: upgrade prefix tree to radix tree

### PR DESCRIPTION
**Pull request by marco-ippolito**
I've created this PR to change the `prefix-tree` with a [`radix-tree`](https://en.wikipedia.org/wiki/Radix_tree) which takes less memory without performance loss
The goal is to have a seamless switch between the two.

To be implemented: 

- [x] exact & tolerance check
- [x]  contains 
- [x] removeDocumentByWord
- [x] removeWord
- [x] test performance

### CODE REVIEW

**Comment by marco-ippolito** [comment]
@micheleriva  any idea on a meaningful benchmark between the two trees?

**Comment by micheleriva** [comment]
@marco-ippolito given the current implementation on `main`, you can run the benchmark suite on both branches. 

cc @RafaelGSS - I'd like to discuss with you any idea on how to make the whole benchmarking thing more accessible when you have time!

**Comment by RafaelGSS** [comment]
> @marco-ippolito given the current implementation on `main`, you can run the benchmark suite on both branches. 
> 
> 
> 
> cc @RafaelGSS - I'd like to discuss with you any idea on how to make the whole benchmarking thing more accessible when you have time!

Sure. Invite to a call.

**Comment by marco-ippolito** [comment]
@micheleriva  is there a reason why the function `removeWord` is never used?
https://github.com/LyraSearch/lyra/blob/85f9e0a3b16798574aa188caf22863098c147647/src/prefix-tree/trie.ts#L150-L173

File: src/radix-tree/radix.ts [hunk 1]
```diff
@@ -0,0 +1,235 @@
...
+export type FindResult = Record<string, string[]>;
+
+export function insert(root: Node, word: string, docId: string) {
+  for (let i = 0; i < word.length; i++) {
+    const currentCharacter = word[i];
+    root.docs.push(docId);
+
+    if (currentCharacter in root.children) {
+      const edgeLabel = root.children[currentCharacter].word;
+      const commonPrefix = getCommonPrefix(edgeLabel, word.substring(i));
```
Comment (ShogunPanda) [inline src/radix-tree/radix.ts:22]: Please extract all repeatedly computed values in variables:

* `root.children[currentCharacter]`
* `word.substring(i)`
* `word.substring(i).length`

and so forth

File: src/radix-tree/radix.ts [hunk 2]
```diff
@@ -0,0 +1,235 @@
...
+        root.children[currentCharacter].docs.push(docId);
+        root.children[currentCharacter].end = true;
+        return;
+      }
+
+      if (commonPrefix.length < edgeLabel.length && commonPrefix.length === word.substring(i).length) {
+        const newNode = createNode(true);
+        newNode.word = word.substring(i);
+
+        newNode.children[edgeLabel[commonPrefix.length]] = root.children[currentCharacter];
```
Comment (micheleriva) [inline src/radix-tree/radix.ts:34]: I guess you can also extract `newNode.children[edgeLabel[commonPrefix.length]]` into a variable here

File: src/radix-tree/radix.ts [hunk 3]
```diff
@@ -0,0 +1,250 @@
...
+      exact,
+      tolerance,
+    );
+  }
+  return output;
+}
+
+function getCommonPrefix(a: string, b: string) {
+  let commonPrefix = "";
+  for (let i = 0; i < Math.min(a.length, b.length); i++) {
```
Comment (micheleriva) [inline src/radix-tree/radix.ts:169]: I would refactor this line from:

```js
  for (let i = 0; i < Math.min(a.length, b.length); i++) {
```

into:

```js
  const len = Math.min(a.length, b.length)

  for (let i = 0; i < len; i++) {
```

**Comment by marco-ippolito** [comment]
It seems the radix tree version runs a little bit slower on my machine
```
## benchmark 0.3.1

╔══════════════════════════════════════════════════════════╤═════════╤══════════════════╤═══════════╗
║ Slower tests                                             │ Samples │           Result │ Tolerance ║
╟──────────────────────────────────────────────────────────┼─────────┼──────────────────┼───────────╢
║ search "decides to leave", tolerance of 1, no stemming   │    2000 │ 451029.98 op/sec │  ± 0.89 % ║
║ search "decides to leave", tolerance of 2, no stemming   │    2500 │ 454022.01 op/sec │  ± 0.85 % ║
╟──────────────────────────────────────────────────────────┼─────────┼──────────────────┼───────────╢
║ Fastest test                                             │ Samples │           Result │ Tolerance ║
╟──────────────────────────────────────────────────────────┼─────────┼──────────────────┼───────────╢
║ search "decides to leave", default settings, no stemming │   10000 │ 456346.15 op/sec │  ± 2.26 % ║
╚══════════════════════════════════════════════════════════╧═════════╧══════════════════╧═══════════╝

╔════════════════════════════════════════════════╤═════════╤══════════════════╤═══════════╗
║ Slower tests                                   │ Samples │           Result │ Tolerance ║
╟────────────────────────────────────────────────┼─────────┼──────────────────┼───────────╢
║ search "beauty", tolerance of 1, no stemming   │    2000 │ 355366.66 op/sec │  ± 1.00 % ║
║ search "beauty", default settings, no stemming │   10000 │ 447471.76 op/sec │  ± 3.40 % ║
╟────────────────────────────────────────────────┼─────────┼──────────────────┼───────────╢
║ Fastest test                                   │ Samples │           Result │ Tolerance ║
╟────────────────────────────────────────────────┼─────────┼──────────────────┼───────────╢
║ search "beauty", tolerance of 2, no stemming   │   10000 │ 448519.09 op/sec │  ± 3.20 % ║
╚════════════════════════════════════════════════╧═════════╧══════════════════╧═══════════╝

╔═════════════════════════════════════════════╤═════════╤══════════════════╤═══════════╗
║ Slower tests                                │ Samples │           Result │ Tolerance ║
╟─────────────────────────────────────────────┼─────────┼──────────────────┼───────────╢
║ search "decides to leave", tolerance of 1   │    2000 │ 454050.95 op/sec │  ± 0.93 % ║
║ search "decides to leave", default settings │   10000 │ 462698.09 op/sec │  ± 2.00 % ║
╟─────────────────────────────────────────────┼─────────┼──────────────────┼───────────╢
║ Fastest test                                │ Samples │           Result │ Tolerance ║
╟─────────────────────────────────────────────┼─────────┼──────────────────┼───────────╢
║ search "decides to leave", tolerance of 2   │   10000 │ 464948.49 op/sec │  ± 2.73 % ║
╚═════════════════════════════════════════════╧═════════╧══════════════════╧═══════════╝

╔═══════════════════════════════════╤═════════╤══════════════════╤═══════════╗
║ Slower tests                      │ Samples │           Result │ Tolerance ║
╟───────────────────────────────────┼─────────┼──────────────────┼───────────╢
║ search "beauty", tolerance of 2   │   10000 │ 394765.54 op/sec │  ± 1.72 % ║
║ search "beauty", default settings │   10000 │ 443211.21 op/sec │  ± 5.91 % ║
╟───────────────────────────────────┼─────────┼──────────────────┼───────────╢
║ Fastest test                      │ Samples │           Result │ Tolerance ║
╟───────────────────────────────────┼─────────┼──────────────────┼───────────╢
║ search "beauty", tolerance of 1   │    2500 │ 443864.62 op/sec │  ± 0.86 % ║
╚═══════════════════════════════════╧═════════╧══════════════════╧═══════════╝

## benchmark radix tree

╔═════════════════════════════════════════════╤═════════╤══════════════════╤═══════════╗
║ Slower tests                                │ Samples │           Result │ Tolerance ║
╟─────────────────────────────────────────────┼─────────┼──────────────────┼───────────╢
║ search "decides to leave", tolerance of 1   │    1500 │ 356540.90 op/sec │  ± 0.97 % ║
║ search "decides to leave", tolerance of 2   │   10000 │ 381497.71 op/sec │  ± 1.83 % ║
╟─────────────────────────────────────────────┼─────────┼──────────────────┼───────────╢
║ Fastest test                                │ Samples │           Result │ Tolerance ║
╟─────────────────────────────────────────────┼─────────┼──────────────────┼───────────╢
║ search "decides to leave", default settings │   10000 │ 469537.93 op/sec │  ± 2.19 % ║
╚═════════════════════════════════════════════╧═════════╧══════════════════╧═══════════╝

╔══════════════════════════════════════════════════════════╤═════════╤══════════════════╤═══════════╗
║ Slower tests                                             │ Samples │           Result │ Tolerance ║
╟──────────────────────────────────────────────────────────┼─────────┼──────────────────┼───────────╢
║ search "decides to leave", default settings, no stemming │   10000 │ 381308.70 op/sec │  ± 1.63 % ║
║ search "decides to leave", tolerance of 1, no stemming   │   10000 │ 426378.07 op/sec │  ± 2.75 % ║
╟──────────────────────────────────────────────────────────┼─────────┼──────────────────┼───────────╢
║ Fastest test                                             │ Samples │           Result │ Tolerance ║
╟──────────────────────────────────────────────────────────┼─────────┼──────────────────┼───────────╢
║ search "decides to leave", tolerance of 2, no stemming   │    2000 │ 452640.05 op/sec │  ± 0.97 % ║
╚══════════════════════════════════════════════════════════╧═════════╧══════════════════╧═══════════╝

╔════════════════════════════════════════════════╤═════════╤══════════════════╤═══════════╗
║ Slower tests                                   │ Samples │           Result │ Tolerance ║
╟────────────────────────────────────────────────┼─────────┼──────────────────┼───────────╢
║ search "beauty", tolerance of 2, no stemming   │   10000 │ 370829.28 op/sec │  ± 1.73 % ║
║ search "beauty", tolerance of 1, no stemming   │   10000 │ 454465.39 op/sec │  ± 2.95 % ║
╟────────────────────────────────────────────────┼─────────┼──────────────────┼───────────╢
║ Fastest test                                   │ Samples │           Result │ Tolerance ║
╟────────────────────────────────────────────────┼─────────┼──────────────────┼───────────╢
║ search "beauty", default settings, no stemming │   10000 │ 471260.40 op/sec │  ± 2.16 % ║
╚════════════════════════════════════════════════╧═════════╧══════════════════╧═══════════╝

╔═══════════════════════════════════╤═════════╤══════════════════╤═══════════╗
║ Slower tests                      │ Samples │           Result │ Tolerance ║
╟───────────────────────────────────┼─────────┼──────────────────┼───────────╢
║ search "beauty", tolerance of 1   │   10000 │ 470995.24 op/sec │  ± 2.06 % ║
║ search "beauty", default settings │   10000 │ 471451.11 op/sec │  ± 1.73 % ║
╟───────────────────────────────────┼─────────┼──────────────────┼───────────╢
║ Fastest test                      │ Samples │           Result │ Tolerance ║
╟───────────────────────────────────┼─────────┼──────────────────┼───────────╢
║ search "beauty", tolerance of 2   │   10000 │ 477426.47 op/sec │  ± 2.19 % ║
╚═══════════════════════════════════╧═════════╧══════════════════╧═══════════╝
```

**Comment by ShogunPanda** [comment]
Amazing! This LGTM!

**Review by ShogunPanda — APPROVED** [review]
LGTM!

**Comment by micheleriva** [comment]
@marco-ippolito, @ShogunPanda I would do the following: I'll merge this to `main`, then try to measure and improve performances before releasing `v0.3.2`.

@marco-ippolito you did awesome work, thank you so much for this!

let me know wdyt

**Comment by ShogunPanda** [comment]
@micheleriva Looks fine to me sir - And anyway you already proceeded! X°°°D

**Pull request merged**